### PR TITLE
[quant][graph] Run dynamic quantization for specific ops

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37125 [quant][graph] Add JIT passes for dynamic quant multi uses of quant node
* **#37093 [quant][graph] Run dynamic quantization for specific ops**
* #37045 [graph][quant] Enable accessing child/grandchild modules in forward
* #37014 [quant][graph] Add check for qconfig_dict key

Summary:
Specify which ops should/can be dynamically quantized. Similar to static quantization

Test Plan:
python test_quantize_script.py test_dynamic_multi_op

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21283695](https://our.internmc.facebook.com/intern/diff/D21283695)

### CODE REVIEW

File: torch/csrc/jit/passes/quantization.cpp [hunk 1]
```diff
@@ -270,13 +278,13 @@ bool mayRequireObservation(Value* v) {
   return !hasScalarInput(v->node());
 }
 
-bool nodeQuantizable(Node* n) {
+bool nodeQuantizable(Node* n, bool is_dynamic) {
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:281]: you can have a default argument for is_dynamic?

File: torch/csrc/jit/passes/quantization.cpp [hunk 2]
```diff
@@ -1101,7 +1109,8 @@ bool InsertObserversHelper::valueNeedsToBeQuantized(Value* v) {
   // of the quantizable function.
   if (!is_dynamic) {
     // Check whether producer is quantizable
-    if (mayRequireObservation(v) && nodeQuantizable(v->node())) {
+    if (mayRequireObservation(v) &&
+        nodeQuantizable(v->node(), /* is_dynamic */ false)) {
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:1113]: so that this can be reverted?

File: torch/csrc/jit/passes/quantization.cpp [hunk 3]
```diff
@@ -53,6 +53,10 @@ std::vector<std::string> _quantizable_call_funcs = {
     "batch_norm",
 };
 
+std::vector<std::string> _quantizable_dynamic_call_funcs = {
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:56]: maybe dynamic_quantizable?

File: torch/csrc/jit/passes/quantization.cpp [hunk 4]
```diff
@@ -53,6 +53,10 @@ std::vector<std::string> _quantizable_call_funcs = {
...
+    "linear",
+};
+
 std::vector<std::string> _quantizable_aten_funcs = {
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:60]: and you can rename this to static_quantizable as well

File: test/quantization/test_quantize_script.py [hunk 5]
```diff
@@ -1739,14 +1739,17 @@ def forward(self, x):
                    .run(m.graph)
 
 
-    def test_insert_quant_dequant_conv_dynamic(self):
+    def test_dynamic_multi_op(self):
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:1742]: why is the name changed? this is supposed to be a unittest for insert_quant_dequant I think, you can add integration tests separately

File: test/quantization/test_quantize_script.py [hunk 6]
```diff
@@ -1739,14 +1739,17 @@ def forward(self, x):
...
                 super(M, self).__init__()
-                self.conv = torch.nn.Conv2d(3, 5, 3).float()
+                self.conv = torch.nn.Conv2d(3, 1, kernel_size=3).to(dtype=torch.float)
+                self.fc1 = torch.nn.Linear(64, 10).to(dtype=torch.float)
 
             def forward(self, x):
-                return self.conv(x)
+                x = self.conv(x)
+                x = x.view(-1, 64).contiguous()
+                return self.fc1(x)
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:1752]: I think it's better to keep this as a unit test

File: test/quantization/test_quantize_script.py (hunk 5 above)
Comment (supriyar) [reply test/quantization/test_quantize_script.py:1742]: There was already a separate test for `test_insert_quant_dequant_linear_dynamic` so this isn't necessary. I'll change it to an e2e test.

File: test/quantization/test_quantize_script.py [hunk 7]
```diff
@@ -1872,6 +1826,25 @@ def forward(self, x):
         FileCheck().check("quantized::linear_dynamic") \
                    .run(model.graph)
 
+    def test_dynamic_multi_op(self):
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:1829]: nit: I think you can add a TestCase for integration tests for dynamic quant

**Review by jerryzh168 — APPROVED** [review]

**Review by jerryzh168 — APPROVED** [review]
Changes looks good, I think we should add a test case for integrration tests for quantize_dynamic_script

**Pull request closed without merging**